File: Data Structures & Algorithms/valid-binary-search-tree/submission-12.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

// Each node key needs to fall between a range
// If we're traversing left the node the node key needs to be smaller than the min val encountered so far
// If we're traversing right the node key needs to be greater than the max val encountered so far
// Time: O(n)
// Space: O(h)

class Solution {
    /**
     * @param {TreeNode} root
     * @return {boolean}
     */
    isValidBST(root: TreeNode | null): boolean {

        const dfs = (node: TreeNode, min: number = -Infinity, max: number = Infinity) => {
            if (!node) return true

            // console.log({val: node.val, min, max})

            if (node.val <= min) return false
            if (node.val >= max) return false

            // dfs(1, -Inifnity, 2)
            const left = dfs(node.left, min, node.val)
            if (!left) return false

            // dfs(3, 2, Inifinty)
            const right = dfs(node.right, node.val, max)
            if (!right) return false

            return true
        }

        return dfs(root)
    }
}

File: Data Structures & Algorithms/climbing-stairs/submission-12.py
class Solution:
    def climbStairs(self, n: int) -> int:
        dp = [None] * (n + 1)

        def dfs(i, path):

            if i == n:
                return 1

            if i > n:
                return 0

            # memoization
            if dp[i] != None:
                return dp[i]

            # dfs(0, 0)
            # LEFT -> dfs(1, 0) -> LEFT dfs(2, 0) + 1 -> RIGHT dfs(4, 0) Return 0
            # RIGHT -> dfs(2, 0) + 1 -> LEFT dfs(1, 0) return 0 -> dfs(4, 0) return 0
            left = dfs(i + 1, path) # 1
            right = dfs(i + 2, path) # 1

            dp[i] = left + right

            return dp[i]

        return dfs(0, 0)





        